Add guidelines for contributing to MewBot

javajawa opened the pull request:
Fixes #9 

- [x] Contributing.md
- [x] code of conduct
- [ ] merge request templates
- [x] set up the contribution bot (let us recognise everyone who ends up helping)

javajawa committed:
Add initial CONTRIBUTING.md

javajawa commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+This document outlines the technical and procedural requirements for your work to be included.
+All contributions, in any form, must also confirm with the [code of conduct](CODE_OF_CONDUCT.md).
+
+> :warning: This document is a living document, and will adapt based on the project's needs.
```
Warning is a bit strong here 
```suggestion
> :information_source: This document is a living document, and will adapt based on the project's needs.
```

javajawa commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+## Scope: What Lives in this Repository?
+
+MewBot's core design includes the tenant of modularity.
```
This implies we have a design philosophy with tenants. We sort of do in the design docs, but it might be cool to codify that some more

javajawa commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+MewBot's core design includes the tenant of modularity.
+The core repo will only accept features and components that do not add additional dependencies.
+
+When building new components, whether they are input/output modules or utilities for use in behaviours,
+consider if they could be a stand-alone repository and package.
+The MewBot team is happy to help set up these repositories in the GitHub group.
```
For now, this should be commented out -- we can deal with splitting up the repo once #13 is solved

javajawa reviewed (commented):
The linter issues will be addressed in a different branch, along with further work on #6 to see if we can solve the underlying issue.

Main bits of feedback for myself so far:
 - This needs examples (please add review comments anywhere people thing it would substantially benefit from an example
 - This PR doesn't fully conform with what I've written. One or the other should be improved until 'theory' matches 'reality'

ajCameron commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+## Scope: What Lives in this Repository?
+
+MewBot's core design includes the tenant of modularity.
+The core repo will only accept features and components that do not add additional dependencies.
```
I mean, once we stabilise what the requirements actually are ...

Definitely aiohttp. Probably some kind of file handling library. Perhaps a decent async database framework for the object stores?

exachixkitsune replied:
I could also see a world where we do accept additional dependencies for isolated cases.

javajawa replied:
I have reworded this section to be positively worded -- it now explains what is part of the core functionality, and that adding third party dependencies requires extra consideration.

ajCameron replied:
I agree with the rewording - both the sentiment and the specific language used.

In practise, I think we're going to struggle with this - especially as we add more io options. If we can keep it down to less than 1 per io type, I'll be personally happy. But there's every reason to have this language in the guide as a target.

This is partially resolvable by a more sophisticated build/install system.

ajCameron commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+Issues are also a place to plan how to approach future work.
+Openly discussing ideas allows us to find places where multiple people are looking at
+similar problems and ensure that we don't have different solutions.
```
Suggest adding something on the lines of "Don't be shy in adding issues. All concerns, or thoughts, are wlecome. But please don't be offended if issues get merged as we work towards a solution - which may solve several of them at once."

ajCameron replied:
Or something on these lines more grammerly acceptable? As opening an issue can be fairly intimidating. I'm thinking of the eventual target audience - who, hopefully, will include quite a lot of low or no code individuals.

ajCameron replied:
The rewording is much better!

ajCameron commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+Openly discussing ideas allows us to find places where multiple people are looking at
+similar problems and ensure that we don't have different solutions.
+
+When starting on a piece of work, check if there is a current issue you can assign to yourself.
```
Suggest adding "please"

ajCameron replied:
But yeah - minor style thing

javajawa replied:
Entirely reworded to a more positive:
> If you want to start on a piece of work: find or create an issue describing the work; assign it to yourself; and get started.

ajCameron commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+Expressing intent is the other key ingredient.
+Name code elements (classes, functions, variables) based on what purpose they serve,
+and try to organise code to convey the flow of logic.
+If there is additional information, use single-line comments.
```
Suggest adding - just as a thought - and very happy to be argued off it. "More is better here in terms of clarity and content. If individual sections of code are becoming unwieldy then extensive comments may be moved into the wiki. If you do this, please include a link to the wiki page section where the comments now reside. Notes as to intent are also welcome."

ajCameron replied:
Ah - see the section below partially deals with it.

exachixkitsune replied:
Would also reccomend adding "Notes to intent should be removed prior to merge, or with a suitable issue raised as part of the pull request process"

ajCameron replied:
That's probably a better solution - that way people can go back and see the notes before the final pruning in the commits that made up the branch? Which is definitely more elegant than leaving them in the final code base.

We could even include links on the relevant wiki page to the last commit containing them.

ajCameron replied:
(Possidbly adding something about how 

"Extensive comments, or comments which should be read prior to working on the component, should be transfered into the wiki and a link to the relevant page included in the code")

ajCameron replied:
(But I am far from fixed on this, and the tone/content of the wiki is a discussion for another point.

Long quite liked the idea of a contributor wiki, where people can dump any notes they make during dev. Will add an issue).

ajCameron commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+Commit messages, like code elements, starts with a summary of the intent of the change,
+and the full version can also include references to the issue(s) it solves and descriptions
+of the functional details of the changes.
+
```
Suggest adding.

"Before you start work on a new feature, please check the issues to reduce the chance of duplication of work. Before you start a new project, please open an issue describing your intent."

Think it bears repeating.

javajawa replied:
I've added it to the top of the review section instead, but yes more repeating

ajCameron commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+When giving feedback of alternative methods or problems, include a clear description and example (pseudo)code or a link to documentation.
+
+It is up to the reviewer (or a maintainer) to mark a conversation as resolved, not the merge requester.
+An exception naturally applies to praise and other comments that do not need a response.
```
Overall well written and clear.

exachixkitsune replied:
Praise: I do like this idea of marking your comments on what we should say.
Potential Optional Addition: do we want to mandate a list/mandate having these tone markers

javajawa replied:
I don't think they _need_ tone markers, per se -- I'm hoping that just having the idea of thesallows for clearer communication (both in the form of people writing comments with language that makes it clearer, and also giving people a way to ask "is this a problem that has to be dealt with?" in a slightly more polite manner

exachixkitsune commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+Issues are a fantastic way of keeping track of what work is needed and who is planning to do it.
+You can find all the current initiatives in the projects tab; these show the current areas of focus.
+If you are looking for a way to contribute to the project, this is a great place to start.
```
```suggestion
If you are looking for a way to contribute to mewbot, this is a great place to start.
```
(Grammar: changing multiple uses of the word "project" in "github projects" and "the mewbot project"

exachixkitsune commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+Issues are also a place to plan how to approach future work.
+Openly discussing ideas allows us to find places where multiple people are looking at
+similar problems and ensure that we don't have different solutions.
```
```suggestion
similar problems and ensure that we don't have different solutions.
Don't be shy in adding issues. All concerns, thoughts or suggestions are welcome. The team may
merge issues as we work towards a common solution, or issues will be dropped if they do not
reflect the overall project vision.
```
Would be my suggestion

exachixkitsune commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+-->
+
+## Branches And Commits
```
Should we invent an image here? Something like:
```
----------------------------------
o (v1.0) - -  o - - - - - - - o (v1.1) (mainline)
 \           /               /
   o - - - o - - - o - - - o               (<issue branch names>)
--------------------------------
```

exachixkitsune commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+and the full version can also include references to the issue(s) it solves and descriptions
+of the functional details of the changes.
+
+## Pull Requests: Getting Stuff Into The Project
```
Do we have enforced merge requirements to prevent errant pull requests being solved by non-maintainers/when all reviewers have not say "yay"

javajawa replied:
Yes; code review requirements are currently:
 - one maintainer approval
 - conversations in the merge resolved (which doesn't mean much because anyone can resolve conversations, iirc)
 - linting and sonar passing

exachixkitsune commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+Contributions to this project are welcome, whether minor updates to documentation or entirely new features.
+This document outlines the technical and procedural requirements for your work to be included.
+All contributions, in any form, must also confirm with the [code of conduct](CODE_OF_CONDUCT.md).
```
TODO: Code of Conduct

exachixkitsune commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+Contributions to this project are welcome, whether minor updates to documentation or entirely new features.
+This document outlines the technical and procedural requirements for your work to be included.
+All contributions, in any form, must also confirm with the [code of conduct](CODE_OF_CONDUCT.md).
```
JOKE: TODO: Code of Resistance. you know, to counteract the conductivity.

ajCameron replied:
heh - did get a chuckle out of that one!

exachixkitsune reviewed (commented):
I'd maybe consider more headings, differentate things into a "Issues" category, a "Code Standards" category, and a "Pull Request/Commentry" Category.

javajawa committed:
Add setup for @all-contributors

This addresses one part of #9.
The initial configuration is based on the automatically generated PR #35

javajawa committed:
Update the contributing document from feedback on #24

javajawa committed:
Add Code of Conduct

The code of conduct selected is the 'Contributor Covenant', v2.
This standardised code of conduct covers our current needs

exachixkitsune reviewed (approved):
All looks good from me with the caveat that I haven't yet read all the license, nor necessarily am aware of it's full implications.
Do we want to add the relevant Creative Commons icon to our readme?

javajawa commented:
> All looks good from me with the caveat that I haven't yet read all the license, nor necessarily am aware of it's full implications. Do we want to add the relevant Creative Commons icon to our readme?

The license has been added as it is the license the code of conduct is released under; it currently only applies to that file, and the main license for the repo is still the BSD 2-Clause. That said #17 addresses the general question of what license we should use and what the copyright holder name should be.

exachixkitsune commented:
> The license has been added as it is the license the code of conduct is released under; it currently only applies to that file, and the main license for the repo is still the BSD 2-Clause. That said #17 addresses the general question of what license we should use and what the copyright holder name should be.

I see. In that case; the CODE_OF_CONDUCT.md doesn't say it's being shared under CC-BY. Should it?

javajawa commented:
> I see. In that case; the CODE_OF_CONDUCT.md doesn't say it's being shared under CC-BY. Should it?

It is, using the [SPDX standard](https://spdx.dev/specifications/)
https://github.com/mewler/mewbot/blob/3c96584/CODE_OF_CONDUCT.md?plain=1#L2

exachixkitsune commented:
> > I see. In that case; the CODE_OF_CONDUCT.md doesn't say it's being shared under CC-BY. Should it?
> 
> It is, using the [SPDX standard](https://spdx.dev/specifications/) https://github.com/mewler/mewbot/blob/3c96584/CODE_OF_CONDUCT.md?plain=1#L2

`>.<` So it is.

ajCameron reviewed (approved):
I believe all issues I raised have been resolved, or, in one case, moved to an issue to consider more generally.

I'm a little worried about the code of conduct being a little intimidating, but there's nothing in  there I don't thing _should_ be in there. So that might not be a solvable problem.
